File: src/utils/neural.js
/**
 * neural.js
 * All neural network and genetic algorithm logic for AVTAS.
 * Completely decoupled from React — pure functions.
 */

/** Deep-clone a neural network weight structure */
export const cloneNetwork = (net) => ({
  weights0: net.weights0.map(row => [...row]),
  biases0:  [...net.biases0],
  weights1: net.weights1.map(row => [...row]),
  biases1:  [...net.biases1],
});

/** Create a randomly initialised network: inputSize→hiddenSize→outputSize */
export const generateRandomNetwork = (inputSize, hiddenSize, outputSize) => {
  const weights0 = Array.from({ length: inputSize  }, () =>
    Array.from({ length: hiddenSize  }, () => Math.random() * 2 - 1)
  );
  const biases0  = Array.from({ length: hiddenSize  }, () => Math.random() * 2 - 1);
  const weights1 = Array.from({ length: hiddenSize  }, () =>
    Array.from({ length: outputSize  }, () => Math.random() * 2 - 1)
  );
  const biases1  = Array.from({ length: outputSize  }, () => Math.random() * 2 - 1);

  return { weights0, biases0, weights1, biases1 };
};

/** Apply gaussian-style mutation to every weight at the given rate/power */
export const mutateNetwork = (net, rate, power = 0.2) => {
  const mut = (v) => Math.random() < rate ? v + (Math.random() * 2 - 1) * power : v;
  return {
    weights0: net.weights0.map(row => row.map(mut)),
    biases0:  net.biases0.map(mut),
    weights1: net.weights1.map(row => row.map(mut)),
    biases1:  net.biases1.map(mut),
  };
};

/** Uniform crossover blend of two parent networks */
export const crossoverNetworks = (parentA, parentB) => {
  const blend = (a, b) => Math.random() > 0.5 ? a : b;
  return {
    weights0: parentA.weights0.map((row, i) => row.map((v, j) => blend(v, parentB.weights0[i][j]))),
    biases0:  parentA.biases0.map((v, i) => blend(v, parentB.biases0[i])),
    weights1: parentA.weights1.map((row, i) => row.map((v, j) => blend(v, parentB.weights1[i][j]))),
    biases1:  parentA.biases1.map((v, i) => blend(v, parentB.biases1[i])),
  };
};

/**
 * Feedforward pass.
 * Inputs: [5 LIDAR sensors, normalised speed]
 * Outputs: [steer, throttle] both in tanh range
 */
export const feedForward = (inputs, net) => {
  const hidden = net.biases0.map((bias, i) => {
    let sum = bias;
    inputs.forEach((inp, j) => { sum += inp * net.weights0[j][i]; });
    return Math.tanh(sum);
  });

  const outputs = net.biases1.map((bias, i) => {
    let sum = bias;
    hidden.forEach((h, j) => { sum += h * net.weights1[j][i]; });
    return Math.tanh(sum);
  });

  return { hidden, outputs };
};

File: backend/neural.py
import numpy as np

class NeuralNetwork:
    def __init__(self, input_size, hidden_size, output_size):
        # Initialize weights with Xavier/Glorot-like uniform distribution
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.output_size = output_size
        
        self.weights0 = np.random.uniform(-1, 1, (input_size, hidden_size))
        self.biases0 = np.random.uniform(-1, 1, hidden_size)
        self.weights1 = np.random.uniform(-1, 1, (hidden_size, output_size))
        self.biases1 = np.random.uniform(-1, 1, output_size)

    def feed_forward(self, inputs):
        """
        Inputs: numpy array of shape (N, input_size) or (input_size,)
        Outputs: numpy array of shape (N, output_size) or (output_size,)
        """
        inputs = np.array(inputs)
        # Hidden layer
        hidden = np.tanh(np.dot(inputs, self.weights0) + self.biases0)
        # Output layer
        outputs = np.tanh(np.dot(hidden, self.weights1) + self.biases1)
        return outputs

    def get_weights(self):
        return {
            "weights0": self.weights0.tolist(),
            "biases0": self.biases0.tolist(),
            "weights1": self.weights1.tolist(),
            "biases1": self.biases1.tolist()
        }

    def set_weights(self, w_dict):
        self.weights0 = np.array(w_dict["weights0"])
        self.biases0 = np.array(w_dict["biases0"])
        self.weights1 = np.array(w_dict["weights1"])
        self.biases1 = np.array(w_dict["biases1"])

    def clone(self):
        new_net = NeuralNetwork(self.input_size, self.hidden_size, self.output_size)
        new_net.set_weights(self.get_weights())
        return new_net
